File: 20validParentheses/validParentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::unordered_map const parentheses{std::pair{'{','}'}, {'[',']'}, {'(',')'}};
        std::stack<char> stack;
        for (auto const _: s)
            if (auto find{parentheses.find(_)}; find != std::cend(parentheses)) stack.emplace(find->second);
            else if (std::empty(stack) || stack.top() != _) return false;
            else stack.pop();
        return std::empty(stack);
    }
};

File: 20validParentheses/validParentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const parentheses = Object.freeze({'{':'}', '[':']', '(':')'})
    stack = []
    for (const _ of s)
        if (_ in parentheses) stack.push(parentheses[_])
        else if (!stack.length || !Object.is(stack.pop(), _)) return false 
    return !stack.length
};
